abr_analyze/nengo_utils/network_utils.py

```diff
@@ -15,490 +15,513 @@
+    ----------
+    n_neurons : int
+        the number of neurons in the simulation, needed to get
+        a corresponding number of encoders
+    input_signal : array(time_steps, dimensions), Optional (Default: None)
+        the input signal to sample from
+    thresh : float, Optional (Default: 0.008)
+        the threshold to keep encoder values minimally apart from
+        It may be helpful to increase this value if there are many more
+        inputs than there are neurons in the final simulation to speed up
+        the encoder selection process
+    depth : int
+        how many times has this function been recursively called
+    """
+
+    # first run so we need to generate encoders for the sessions
+    ii = 0
+    iters_with_no_update = 0
+    prev_n_indices = 0
+    while input_signal.shape[0] > n_neurons:
+
+        ii += 1
+        if (ii % 1000) == 0:
+            print('Downsampled to %i encoders at iteration %i' %
+                  (input_signal.shape[0], ii))
+            print('Current threshold value: %.3f' % thresh)
+
+        # choose a random set of indices
+        n_indices = input_signal.shape[0]
+
+        # make sure we're dealing with an even number
+        n_indices -= int(n_indices % 2)
+        n_half = int(n_indices / 2)
+
+        # split data into two random groups
+        randomized_indices = np.random.permutation(range(n_indices))
+        a = randomized_indices[:n_half]
+        b = randomized_indices[n_half:]
+        data1 = input_signal[a]
+        data2 = input_signal[b]
+
+        # calculate the 2 norm between random pairs between data1 and 2
+        distances = np.linalg.norm(data1 - data2, axis=1)
+        # find any pairs within threshold distance of each other
+        under_thresh = distances > thresh
+        # remove the values from data2 within thresh of corresponding data1
+        input_signal = np.vstack([data1, data2[under_thresh]])
+
+        if prev_n_indices == n_indices:
+            iters_with_no_update += 1
+        else:
+            iters_with_no_update = 0
+
+        # if we've run 50 iterations but we're still haven't downsampled
+        # enough, increase the threshold distance and keep going
+        if iters_with_no_update == 50:
+            iters_with_no_update = 0
+            thresh += .1 * thresh
+            print('All values within threshold, but not at target size.')
+            print('Increasing threshold to %.4f' % thresh)
+        prev_n_indices = n_indices
+
+    if input_signal.shape[0] != n_neurons:
+        print('Too many indices removed, appending samples from ' +
+              'ScatteredHypersphere')
+        length = n_neurons - input_signal.shape[0] + 1
+        hypersphere = ScatteredHypersphere(surface=True)
+        hyper_inputs = hypersphere.sample(length, input_signal.shape[1])
+        input_signal = np.vstack((input_signal, hyper_inputs))
+
+        # make sure that the new inputs meet threshold constraints
+        if depth < 10:
```

Comment:
when points are added from scattered hypersphere, want to rerun the threshold testing to avoid adding any points inside threshold distance to other points